File: src/Read2Me.Core/Models/AliasCollisions.cs
namespace Read2Me.Core.Models
{
    /// <summary>
    /// One owner of identity strings in a collision check: a discovery row about to be applied, or
    /// a character already on the roster. <see cref="ExistingCharacterId"/> on a row names the roster
    /// character it will resolve onto, so that character is folded into the row rather than counted
    /// as a second owner.
    /// </summary>
    public sealed record AliasClaim(string Name, IReadOnlyList<string> Aliases, Guid? ExistingCharacterId = null);

    /// <summary>A roster character in a collision check: its id, primary name and aliases.</summary>
    public sealed record AliasOwner(Guid Id, string Name, IReadOnlyList<string> Aliases);

    /// <summary>
    /// Finds identity strings — a character's name or one of its aliases — that would end up
    /// claimed by more than one character once the discovery rows are applied.
    /// </summary>
    /// <remarks>
    /// Attribution resolves a speaker string to a character by first match over an alphabetically
    /// ordered roster (<c>CharacterResolver.ResolveOrCreateAsync</c>), so a string owned by two
    /// characters silently binds to whichever sorts first, in every scene. Discovery is where these
    /// arrive: asked for the cast of <i>Pride and Prejudice</i>, the LLM handed <c>Miss Bennet</c> to
    /// all five Bennet daughters. This is advisory — the review surfaces it and the user removes the
    /// offending alias. Nothing here blocks an apply. The web client carries a line-for-line port
    /// (<c>alias-collisions.ts</c>) tested against the same fixtures.
    /// </remarks>
    public static class AliasCollisions
    {
        /// <summary>
        /// Returns the strings owned by two or more characters, comparing case-insensitively.
        /// Pass only the rows that will be applied — an excluded row is never applied.
        /// </summary>
        public static IReadOnlySet<string> Find(IEnumerable<AliasClaim> rows, IEnumerable<AliasOwner> roster)
        {
            var included = rows.ToList();
            var claimed = included
                .Where(r => r.ExistingCharacterId is { } id)
                .Select(r => r.ExistingCharacterId!.Value)
                .ToHashSet();

            var owners = included
                .Select(r => r.Aliases.Prepend(r.Name))
                .Concat(roster
                    .Where(c => !claimed.Contains(c.Id))
                    .Select(c => c.Aliases.Prepend(c.Name)));

            var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            var collisions = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            foreach (var owner in owners)
            {
                // Distinct within an owner: one character listing the same alias twice is untidy,
                // not ambiguous.
                foreach (var name in owner
                             .Where(n => !string.IsNullOrWhiteSpace(n))
                             .Select(n => n.Trim())
                             .Distinct(StringComparer.OrdinalIgnoreCase))
                {
                    if (!seen.Add(name))
                        collisions.Add(name);
                }
            }
            return collisions;
        }
    }
}

File: src/Read2Me.Web/src/app/pages/cast/alias-collisions.ts
/**
 * Line-for-line port of `Read2Me.Core.Models.AliasCollisions`, tested against the same fixtures
 * (`alias-collisions.spec.ts` ⇔ `AliasCollisionsTests.cs`). The host answers the initial set with
 * the discovery outcome; the review dialog recomputes it as the user edits rows.
 *
 * Attribution resolves a speaker string to one character by first match over the roster, so a
 * string two characters own silently binds to whichever sorts first. Advisory: nothing here blocks
 * an apply.
 */

/** A discovery row about to be applied; `existingCharacterId` names the roster character it folds into. */
export interface AliasClaim {
  name: string;
  aliases: readonly string[];
  existingCharacterId?: string | null;
}

/** A roster character: id, primary name and aliases. */
export interface AliasOwner {
  id: string;
  name: string;
  aliases: readonly string[];
}

/**
 * The strings owned by two or more characters, in first-seen spelling; compare with
 * {@link collides}. Pass only the rows that will be applied.
 */
export function findAliasCollisions(
  rows: readonly AliasClaim[],
  roster: readonly AliasOwner[],
): string[] {
  const claimed = new Set(
    rows.map((r) => r.existingCharacterId).filter((id): id is string => !!id),
  );
  const owners: (readonly string[])[] = [
    ...rows.map((r) => [r.name, ...r.aliases]),
    ...roster.filter((c) => !claimed.has(c.id)).map((c) => [c.name, ...c.aliases]),
  ];

  const seen = new Set<string>();
  const collisions = new Map<string, string>();
  for (const owner of owners) {
    // Distinct within an owner: one character listing the same alias twice is untidy, not ambiguous.
    const names = new Map<string, string>();
    for (const raw of owner) {
      const name = raw.trim();
      if (name.length === 0) continue;
      const key = fold(name);
      if (!names.has(key)) names.set(key, name);
    }
    for (const [key, name] of names) {
      if (seen.has(key)) {
        if (!collisions.has(key)) collisions.set(key, name);
      } else {
        seen.add(key);
      }
    }
  }
  return [...collisions.values()];
}

/** Whether `name` is one of the collisions, case- and whitespace-insensitively. */
export function collides(collisions: readonly string[], name: string): boolean {
  const key = fold(name.trim());
  return collisions.some((c) => fold(c) === key);
}

function fold(name: string): string {
  return name.toLowerCase();
}
